deflate_p.h

```diff
@@ -27,21 +29,39 @@ void flush_pending(PREFIX3(stream) *strm);
  *    (except for the last MIN_MATCH-1 bytes of the input file).
  */
 
-static inline Pos insert_string_c(deflate_state *const s, const Pos str, unsigned int count) {
-    Pos ret = 0;
-    unsigned int idx;
-
-    for (idx = 0; idx < count; idx++) {
-        UPDATE_HASH(s, s->ins_h, str+idx);
-
-        Pos head = s->head[s->ins_h];
-        if (head != str+idx) {
-          s->prev[(str+idx) & s->w_mask] = head;
-          s->head[s->ins_h] = str+idx;
-          if (idx == count - 1)
-            ret = head;
-        } else if (idx == count - 1) {
-          ret = str + idx;
+static inline Pos insert_string(deflate_state *const s, const Pos str, unsigned int count) {
+    Pos str_idx, str_end, ret;
+
+    if (UNLIKELY(count == 0)) {
+        return s->prev[str & s->w_mask];
+    }
+
+    ret = 0;
+    str_idx = str;
+    str_end = str + count - 1; /* last position */
+
+    for (str_idx = str; str_idx <= str_end; str_idx++) {
+        uint32_t val, hm;
+#if defined(UNALIGNED_OK)
+        val = *(unsigned int *)(&s->window[str_idx]);
+#else
+        memcpy(&val, &s->window[str_idx], sizeof(val));
+#endif
+
+        if (s->level >= TRIGGER_LEVEL)
+            val &= 0xFFFFFF;
+
+        s->ins_h = functable.update_hash(s, s->ins_h, val);
```

Comment:
It might be faster to store function pointer from ```functable.update_hash``` in local variable before the loop, as then compiler can optimize it as simple jump using register. Some compilers might assume global struct members are volatile by default and as such require indirect jump.

Reply:
Fixed